packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts

```diff
@@ -33,10 +33,449 @@ export interface IVirtualNode extends cdk.IResource {
+
+  constructor(props?: Http2NodeListenerProps) {
+    super();
+    const checkedProps = props ?? {};
+    this.port = checkedProps.port ? checkedProps.port : 8080;
+    this.healthCheck = checkedProps.healthCheck;
+    this.timeout = checkedProps.timeout;
+  }
+
+  /**
+   * Return Listener for HTTP2 protocol when Listener is added to Virtual Node.
+   */
+  public bind(_scope: cdk.Construct): CfnVirtualNode.ListenerProperty {
+    return {
+      portMapping: {
+        port: this.port,
+        protocol: Protocol.HTTP2,
+      },
+      healthCheck: Http2NodeListener.renderHealthCheck({
+        port: this.port,
+        protocol: Protocol.HTTP2,
+      }, this.healthCheck),
+      timeout: this.timeout ? Http2NodeListener.renderTimeout(this.timeout) : undefined,
+    };
+  }
+}
+
+/**
+ * Represents the properties required to define a GRPC Listener for a VirtualNode.
+ */
+export class GrpcNodeListener extends VirtualNodeListener {
+/**
+ * Returns the ListenerTimeoutProperty for GRPC protocol
+ */
+  public static renderTimeout(tm: GrpcTimeout): CfnVirtualNode.ListenerTimeoutProperty {
+    return ({
+      grpc: {
+        idle: tm?.idle !== undefined ? {
+          unit: 'ms',
+          value: tm?.idle.toMilliseconds(),
+        } : undefined,
+        perRequest: tm?.perRequest !== undefined ? {
+          unit: 'ms',
+          value: tm?.perRequest.toMilliseconds(),
+        } : undefined,
+      },
+    });
+  }
+
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
+
+  /**
+   * Listener timeout property for GRPC protocol
+   *
+   * @default - none
+   */
+  readonly timeout?: GrpcTimeout;
+
+  constructor(props?: GrpcNodeListenerProps) {
+    super();
+    const checkedProps = props ?? {};
+    this.port = checkedProps.port ? checkedProps.port : 8080;
+    this.healthCheck = checkedProps.healthCheck;
+    this.timeout = checkedProps.timeout;
+  }
+
+  /**
+   * Return Listener for GRPC protocol when Listener is added to Virtual Node.
+   */
+  public bind(_scope: cdk.Construct): CfnVirtualNode.ListenerProperty {
+    return {
+      portMapping: {
+        port: this.port,
+        protocol: Protocol.GRPC,
+      },
+      healthCheck: GrpcNodeListener.renderHealthCheck({
+        port: this.port,
+        protocol: Protocol.GRPC,
+      }, this.healthCheck),
+      timeout: this.timeout ? GrpcNodeListener.renderTimeout(this.timeout) : undefined,
+    };
+  }
+}
+
+/**
+ * Represents the properties required to define a TCP Listener for a VirtualNode.
+ */
+export class TcpNodeListener extends VirtualNodeListener {
+/**
+ * Returns the ListenerTimeoutProperty for TCP protocol
+ */
+  public static renderTimeout(tm: TcpTimeout): CfnVirtualNode.ListenerTimeoutProperty | undefined {
```

Comment:
Same here - should be a utility, not a public API of this class.

(BTW, there's a ton of duplication in these. You might want to think about DRYing this whole `VirtualNodeListener` hierarchy up)